TypeScript based documentation generation

petebacondarwin opened the pull request:
Closes #1729

petebacondarwin committed:
docs(directives.js): remove invalid asterisk

This character was breaking the doc-gen

petebacondarwin committed:
docs(di/injector): remove invalid tab chars

These tabs were breaking the jade syntax generation

petebacondarwin committed:
chore(doc-gen): fix path to modules on angular.io

petebacondarwin committed:
chore(doc-gen): move inline link matching to dgeni-package

petebacondarwin committed:
chore(doc-gen): remove Traceur bits

petebacondarwin committed:
chore(doc-gen): add TypeScript parsing

petebacondarwin committed:
chore(doc-gen): relax link matching

petebacondarwin commented:
Right! This PR is cleaned up and ready to be reviewed/merged.

There are a number of commits but each should be fairly self-explanatory and independent of the others.

Highlights of this PR are:

* No more dependency on Traceur
* All code components are now parsed using the TypeScript API
* Re-exported components are now resolved automatically - no more need for `@exportedAs` tags. These should be removed from the codebase.
* The "public" modules to publish in the docs are now specified in `readTypeScriptModules.sourceFiles`. Only these modules will appear but code that is imported from other "private" modules are resolved automatically.
* The function signatures of methods, constructors and module level functions are now cleaner and include the return type if specified.
* All module and class pages now contain a valid link to their original Github URL.
* As there are a number of ambiguous links to be fixed, the link parsing is now relaxed to allow the build while provided comprehensive informative warnings.

There seems to be some inconsistency in the code right now where components are "reexported" to components that do not actually exist: `angular/pipes` for example. Either this module should actually export something or be removed.

vicb commented on docs/dgeni-package/index.js:
```diff
@@ -54,19 +46,23 @@ module.exports = new Package('angular', [jsdocPackage, nunjucksPackage])
   ];
+
+  readTypeScriptModules.sourceFiles = [
+    '*/*.js',
```
is `*/*.[js,ts,es6]` syntax supported ?

Not a big deal, just being curious.

petebacondarwin replied:
Probably since it uses glob

petebacondarwin replied:
No, it appears that this is not a valid syntax

petebacondarwin replied:
But this is:

```js
    '*/*.@(js|es6|ts])',
    '*/src/**/*.@(js|es6|ts)'
```

vicb replied:
seems like {js,ts,es6} should be

On 05/15/2015 05:44 PM, Pete Bacon Darwin wrote:
>
> In docs/dgeni-package/index.js 
> <https://github.com/angular/angular/pull/1886#discussion_r30419758>:
>
> >      { include: 'modules/*/docs/**/*.md', basePath: 'modules' },
> >      { include: 'docs/content/**/*.md', basePath: 'docs/content' }
> >    ];
> > +
> > +  readTypeScriptModules.sourceFiles = [
> > +    '*/*.js',
>
> No, it appears that this is not a valid syntax
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/angular/angular/pull/1886/files#r30419758>.
>

petebacondarwin replied:
https://github.com/isaacs/node-glob#glob-primer

vicb commented on docs/dgeni-package/mocks/testSrc.ts:
```diff
@@ -0,0 +1,38 @@
+/**
+ * An exported function
+ */
+export var myFn = (val:number) => { return val*2; }
```
`(val:number) => val*2; `

petebacondarwin replied:
Right. Thanks

vicb commented on docs/dgeni-package/services/tsParser/getContent.js:
```diff
@@ -0,0 +1,49 @@
+          content += sourceFile.text
+              .substring(commentRange.pos+ '/**'.length, commentRange.end - '*/'.length)
+              .replace(LEADING_STAR, '')
+              .trim()
```
missing ";"

petebacondarwin replied:
Thanks

vicb commented:
@petebacondarwin I'm sorry I can not make smarter comments, I don't know the code well enough...
But from your description above, it sounds great !

petebacondarwin committed:
chore(doc-gen): add TypeScript parsing

petebacondarwin committed:
chore(doc-gen): relax link matching

petebacondarwin committed:
chore(doc-gen): relax link matching

petebacondarwin committed:
chore(doc-gen): relax link matching

petebacondarwin committed:
chore(doc-gen): chore(doc-gen): add TypeScript parsing

IgorMinar commented:
I'm looking into the build error

IgorMinar commented on docs/dgeni-package/services/tsParser.js:
```diff
@@ -0,0 +1,68 @@
+    parse: function(fileNames, baseDir) {
+      // This is the easiest way I could find to ensure that we loaded
+      // modules with paths relative to the baseDir
+      process.chdir(baseDir);
```
the build is broken because of this line.

if you change the currentDirectory, it affects everything in the build and since gulp runs a ton of stuff in a single node instance in parallel you actually don't even know what you are affecting and when. this leads to race conditions that break the build.

there are several solutions:
- the easiest one is to change `test._js.sh` to run docs tasks in a separate gulp process. change `gulp build.js docs` to `gulp build.js && gulp docs`
- better solution would be to not change the working directory since it can break many other things
- best solution would be to change our build system completely and run tasks that don't depend on each other in a separate process

petebacondarwin replied:
Thanks for identifying this @IgorMinar. I'll take another look at removing this line Over the weekend.

petebacondarwin replied:
Actually now that I ended up hacking my own CompilerHost it might be possible to do the path resolution there.

petebacondarwin committed:
squash-me: remove unnecessary (and very naughty) `chdir()` call

mprobst commented on docs/dgeni-package/processors/readTypeScriptModules.js:
```diff
@@ -0,0 +1,201 @@
+        moduleSymbol.exportArray.forEach(function(exportSymbol) {
+
+          // Ignore exports starting with an underscore
+          if (hideSpecialExports && exportSymbol.name.charAt(0) === '_') return;
```
isn't this the same as "private", essentially?

mprobst replied:
rename `hideSpecialExport` to something more explanatory? `hideUnderbarExports` or so?

petebacondarwin replied:
I think this is just roll over from the early days. In TypeScript I guess we can always, now, apply a private modifier to a method to make it private so that we don't have to rely on underscores no more. I will refactor this a bit.

petebacondarwin replied:
Oh, actually, I just remembered that this is to catch this `__esModule` hack. Module exports are, by definition, public. I will make this more clear.

mprobst commented on docs/dgeni-package/processors/readTypeScriptModules.js:
```diff
@@ -0,0 +1,201 @@
+              var memberSymbol = resolvedExport.members[memberName];
+              var memberDoc = createMemberDoc(memberSymbol, exportDoc, basePath, parseInfo.typeChecker);
+
+              // We special case the constructor and sort the other members alphabetically
```
Not sure I agree - wouldn't authors usually have a point in putting some methods first, some later? If it's for findability, an alphabetically sorted TOC might be better, don't you think

petebacondarwin replied:
This is a classic debate (like tabs vs spaces). I am agnostic but I believe it was implemented this way because that is what people asked for.

I will make this configurable so that it is easy to change in the future.

petebacondarwin replied:
Done

mprobst commented on docs/dgeni-package/processors/readTypeScriptModules.js:
```diff
@@ -0,0 +1,201 @@
+    var sourceFile = ts.getSourceFileOfNode(declaration);
+    if(!declaration.parameters) {
+      console.log(declaration);
+      throw 'missing declaration parameters';
```
You can get source/line/col information from the declaration, could you include that in the error you throw? That'd make tracking down issues much easier.

Also, please consider throwing `Error`, so we can get stack traces.

petebacondarwin replied:
Right! This was mostly for me when I was debugging the implementation, it should not actually happen in practice - i.e. it would be a programming error on my part or a change to the TypeScript API.
I will make the error better.

petebacondarwin replied:
Done: 31a2172d7f9cfdd1a59f9e6dcae20ffa433667d8

mprobst commented on docs/dgeni-package/processors/readTypeScriptModules.js:
```diff
@@ -0,0 +1,201 @@
+      console.log(declaration);
+      throw 'missing declaration parameters';
+    }
+    var signature = typeChecker.getSignatureFromDeclaration(declaration);
```
signature is unused?

petebacondarwin replied:
Again a leftover from the many iterations of trying just about every possible combination of API paths to get the data I needed. I will remove.

petebacondarwin replied:
Done c36132b5626e13a0ccb9b05c1aeb45efc608687f

mprobst commented on docs/dgeni-package/processors/readTypeScriptModules.js:
```diff
@@ -0,0 +1,201 @@
+    var declaration = symbol.valueDeclaration || symbol.declarations[0];
+    var sourceFile = ts.getSourceFileOfNode(declaration);
+    if(declaration.type) {
+      var signature = typeChecker.getSignatureFromDeclaration(declaration);
```
unused signature again?

petebacondarwin replied:
On it.

petebacondarwin replied:
Done c36132b5626e13a0ccb9b05c1aeb45efc608687f

mprobst commented on docs/dgeni-package/processors/readTypeScriptModules.js:
```diff
@@ -0,0 +1,201 @@
+  function getReturnType(typeChecker, symbol) {
+    var declaration = symbol.valueDeclaration || symbol.declarations[0];
+    var sourceFile = ts.getSourceFileOfNode(declaration);
+    if(declaration.type) {
```
ws after "if_(" missing.

petebacondarwin replied:
Gotcha!

petebacondarwin replied:
Done 301e51917ba6d8a4bea8093c5e57c1df43e5c7bb

petebacondarwin committed:
chore(doc-gen): clarify and improve matching of exports to ignore

Previously this was a bit cryptic as we just had a `hideSpecialExports` flag.
Now it ignores exports that match an array of regexes, with the default case
being those exports called `___esModule`.

mprobst commented on docs/dgeni-package/processors/readTypeScriptModules.js:
```diff
@@ -0,0 +1,201 @@
+
+};
+
+function insertSorted(collection, item, property) {
```
I don't think it matters much, but this is `O(n^2)` insertion sort. Have you considered just collecting all the items and then using `items.sort(function(a, b) { return a.localeCompare(b); });`? Might need a smarter comparator, and might need to first cut off the ctor, then glue it back on in front, or something like that.

petebacondarwin replied:
Yes, you're right but I think that it doesn't matter much in the bigger picture. The biggest bottleneck in dgeni is the file access.

petebacondarwin replied:
I updated this to a better sort: 0291ae7dd4be40cc4bd7f03f8dc151cb9e73f675
The `localeCompare` version was problematic.

Also I remembered that the reason I did it this way, previously, was that in the old days you added members in at other points in the processing, so it was much harder to find a point where you could say: "OK now I have all the members so I can now sort them".

mprobst commented on docs/dgeni-package/index.js:
```diff
@@ -74,6 +66,20 @@ module.exports = new Package('angular', [jsdocPackage, nunjucksPackage])
   parseTagsProcessor.tagDefinitions.push(require('./tag-defs/private'));
   parseTagsProcessor.tagDefinitions.push(require('./tag-defs/exportedAs'));
+
+  // We actually don't want to parse param docs in this package as we are getting the data out using TS
```
probably just my ignorance on how this works, but wouldn't we still want to read `@param` tags for the parameter description?

petebacondarwin replied:
This is because, "right now" the docs don't display any info about the params. This is extreme programming at its best :-)

mprobst commented on docs/dgeni-package/services/tsParser.js:
```diff
@@ -0,0 +1,69 @@
+    },
+
+    parse: function(fileNames, baseDir) {
+      fileNames = fileNames.map(function(fileName) {
```
Generally speaking it's often nicer to *not* create absolute paths, but rather keep the relative path name as that makes for much more readable error messages.

As you're creating your own compiler host below anyway, maybe you could move the `basePath` handling there?

petebacondarwin replied:
We could definitely do the resolve inside the compiler host but `getSourceFile` gets hit many many times by the compiler and I think this would be wasteful. The `fileNames` array is only used for this first set of files for which we want to create docs. Inside the docs we have access to both the absolute and the relative file path anyway, so error messages are still readable.

petebacondarwin replied:
I had a play with moving the `baseDir` into the compiler host but this totally messes up the parsing, when running `gulp docs/angular.io`, since you get some symbols that don't appear to contain a declaration - which I suspect is due to not matching the symbol back to the originating file.
It doesn't seem worth spending the extra development time to work out what is going on there.

mprobst commented on docs/dgeni-package/services/tsParser.js:
```diff
@@ -0,0 +1,69 @@
+        // by the compiler/binder, it seems that explicit imports (e.g. `export {SomeClass} from 'some/module'`)
+        // do not so we have to do a little work.
+        tsModule.exportArray.forEach(function(moduleExport) {
+          if (moduleExport.flags & 8388608 /* Alias */) {
```
Can you explain the magic constant and/or maybe file a bug against TypeScript that you have to do something like this?

petebacondarwin replied:
This is a TypeScript enum value. The comment next to it was supposed to explain this.
Since writing this piece of code, I now know that I can get this enum directly as `ts.SymbolFlags.Alias` so I will change this.
Thanks!

petebacondarwin replied:
Done 5ff2077ca2086ac96ef58ba26f838b9856dbd93e

mprobst commented:
LGTM for the TypeScript handling parts, modulo the comments about file handling, would be nice to fix that.

petebacondarwin commented:
@mprobst - thank you for this excellent and thorough review. I am working through the changes need (and adding more unit tests).

petebacondarwin committed:
chore(doc-gen): allow class member sorting to be configured

petebacondarwin committed:
chore(doc-gen): improve error handling for missing declaration params

petebacondarwin committed:
chore(doc-gen): remove unused `signature`

petebacondarwin committed:
chore(doc-gen): formatting

petebacondarwin committed:
chore(doc-gen): speed up class member sorting

petebacondarwin committed:
chore(doc-gen): use actual TypeScript SymbolFlags enum rather than magic number

petebacondarwin committed:
chore(doc-gen): speed up class member sorting

eggers commented on docs/angular.io-package/templates/function.template.html:
```diff
@@ -5,10 +5,15 @@
   p.location-badge.
-    exported from <a href="/{$ doc.moduleDoc.path $}">{$ doc.moduleDoc.id $}</a>
+    exported from {@link {$ doc.moduleDoc.id $} {$doc.moduleDoc.id $} }
+    defined in <a href="https://github.com/angular/angular/tree/master/modules/{$ doc.fileInfo.relativePath $}#L{$ doc.location.start.line+1 $}-L{$ doc.location.end.line+1 $}">{$ doc.fileInfo.relativePath $} (line {$ doc.location.start.line+1 $})</a>
```
It would be nice, if instead of using `master`, the actual commit hash were used. This is mainly an issue with the current status of files changing far more frequently than the angular.io site getting updated. It could also become more of an issue when people are using older versions of the docs, and angular.io has docs available for multiple versions (as is currently available on the code.angularjs.org site.

naomiblack replied:
@eggers I like this -- could you submit it as a feature request (or even a PR)? This is blocking a release, so I'd like to get it in and then tweak further in subsequent updates.

eggers replied:
Absolutely, don't want to hold up the good work on this PR

petebacondarwin replied:
@naomiblack and @eggers I have put in a PR for this at #1994. Can you review?

naomiblack commented:
I tried patching this and then running gulp, I get the following error: 

```
gulp public_docs/angular.io
Dart SDK detected
[17:25:43] Using gulpfile ~/github/angular/gulpfile.js
[17:25:43] Task 'public_docs/angular.io' is not in your gulpfile
```
@alexwolfe can you generate the site docs? @petebacondarwin anything obvious I am missing?

naomiblack commented:
This is looking good to merge. @alexwolfe @petebacondarwin you can see the API docs on angular.io here https://github.com/angular/angular.io/pull/121 (you can check out my branch and run on your own server)

eggers commented:
I'm having a problem with the basePath not getting stripped off when I run `gulp docs/angular.io`

It's dumping the partials into `dist/angular.io/partials/api/{basePath}/angular2`, Also, all of the links have the `basePath` sandwiched in there.

Actual: 
https://github.com/angular/angular/tree/master/modules/{basePath}/angular2/src/change_detection/change_detection.ts#L118-L135

Should Be:
https://github.com/angular/angular/blob/master/modules/angular2/src/change_detection/change_detection.ts#L118-L135

Any idea if I'm doing something wrong?

petebacondarwin commented:
@eggers : Yes, I just saw that too. I am looking into it now

petebacondarwin committed:
chore(doc-gen): fix text mock code

petebacondarwin committed:
chore(doc-gen): get CompilerHost to return the SourceFiles with the correct fileName

Now that we are not changing the working directory, we needed to update how
getSourceFiles works to ensure that it worked relative to the baseDir.

petebacondarwin committed:
chore(doc-gen): include interfaces in the possible exports from a module

petebacondarwin commented:
@eggers this should be fixed now.

petebacondarwin committed:
chore(doc-gen): paths now start with a slash

petebacondarwin committed:
chore(doc-gen): fix paths for links to modules and exports

petebacondarwin committed:
chore(doc-gen): fix up tests to work on linux

petebacondarwin commented:
Merged into master.

eggers commented:
@petebacondarwin what master has this been merged into?

petebacondarwin commented:
The one on this repo
